src/components/implementation/cbuf_mgr/default/cbuf_mgr.c

```diff
@@ -0,0 +1,823 @@
+	return &cmr->m[cbid - cmr->low_id];
+}
+
+static struct cbuf_meta_range *
+cbuf_meta_add(struct cbuf_comp_info *comp, u32_t cbid, struct cbuf_meta *m, vaddr_t dest)
+{
+	struct cbuf_meta_range *cmr;
+
+	if (cbuf_meta_lookup(comp, cbid)) return NULL;
+	cmr = malloc(sizeof(struct cbuf_meta_range));
+	if (!cmr) return NULL;
+	INIT_LIST(cmr, next, prev);
+	cmr->m      = m;
+	cmr->dest   = dest;
+	/* must be power of 2: */
+	cmr->low_id = (cbid & ~((PAGE_SIZE/sizeof(struct cbuf_meta))-1));
+
+	if (comp->cbuf_metas) ADD_LIST(comp->cbuf_metas, cmr, next, prev);
+	else                  comp->cbuf_metas = cmr;
+
+	return cmr;
+}
+
+static void
+cbuf_comp_info_init(spdid_t spdid, struct cbuf_comp_info *cci)
+{
+	memset(cci, 0, sizeof(*cci));
+	cci->spdid = spdid;
+	INIT_LIST(&cci->bthd_list, next, prev);
+	cci->limit_size = INIT_LIMIT_SIZE;
+	cvect_add(&components, cci, spdid);
+}
+
+static struct cbuf_comp_info *
+cbuf_comp_info_get(spdid_t spdid)
+{
+	struct cbuf_comp_info *cci;
+
+	cci = cvect_lookup(&components, spdid);
+	if (!cci) {
+		cci = malloc(sizeof(*cci));
+		if (!cci) return NULL;
+		cbuf_comp_info_init(spdid, cci);
+	}
+	return cci;
+}
+
+static struct cbuf_bin *
+cbuf_comp_info_bin_get(struct cbuf_comp_info *cci, int sz)
+{
+	int i;
+
+	assert(sz);
+	for (i = 0 ; i < cci->nbin ; i++) {
+		if (sz == cci->cbufs[i].size) return &cci->cbufs[i];
+	}
+	return NULL;
+}
+
+static struct cbuf_bin *
+cbuf_comp_info_bin_add(struct cbuf_comp_info *cci, int sz)
+{
+	if (sz == CBUF_MAX_NSZ) return NULL;
+	cci->cbufs[cci->nbin].size = sz;
+	cci->nbin++;
+
+	return &cci->cbufs[cci->nbin-1];
+}
+
+static int
+cbuf_map(spdid_t spdid, vaddr_t daddr, void *page, int size, int flags)
+{
+	int off;
+	assert(size == (int)round_to_page(size));
+	assert(daddr);
+	assert(page);
+	for (off = 0 ; off < size ; off += PAGE_SIZE) {
+		vaddr_t d = daddr + off;
+		if (d != (mman_alias_page(cos_spd_id(), ((vaddr_t)page) + off,
+						spdid, d, flags))) {
+			assert(0); /* TODO: roll back the aliases, etc... */
+		}
+	}
+	return 0;
+}
+
+static int
+cbuf_alloc_map(spdid_t spdid, vaddr_t *daddr, void **page, int size)
+{
+	void *p;
+	vaddr_t dest;
+	int ret;
+
+	assert(size == (int)round_to_page(size));
+	p = page_alloc(size/PAGE_SIZE);
+	assert(p);
+	memset(p, 0, size);
+
+	dest = (vaddr_t)valloc_alloc(cos_spd_id(), spdid, size/PAGE_SIZE);
+	assert(dest);
+
+	ret = cbuf_map(spdid, dest, p, size, MAPPING_RW);
+	if (ret) valloc_free(cos_spd_id(), spdid, (void *)daddr, 1);
+	*page  = p;
+	*daddr = dest;
+
+	return ret;
+}
+
+static inline void
+cbuf_add_to_blk_list(struct cbuf_comp_info *cci, int request_size)
+{
+	struct blocked_thd *bthd;
+	bthd = malloc(sizeof(struct blocked_thd));
+	if (unlikely(bthd == NULL)) BUG();
+	bthd->thd_id = cos_get_thd_id();
+	//bthd->spdid = cci->spdid;
+	bthd->request_size = request_size;
+	ADD_LIST(&cci->bthd_list, bthd, next, prev);
+	++cci->num_blocked_thds;
+}
+static int
+cbuf_wake_up_thd(struct cbuf_comp_info *cci)
```

Comment:
I'll pull this as the manager changes are quite isolated, but I you to figure out a way to do a review later.  Maybe it will be a code walk through with Qi and Gedare in the room.